File: 2022/adventofcode/day01.py
from aoc import load_day


__day__ = 1


def parse_input(input: str):
    return list(
        list(int(each.strip()) for each in group.strip().split("\n")) for group in input
    )


def part1(data: list[list[int]]) -> int:
    return max(map(sum, data))


def part2(data: list[list[int]]) -> int:
    return sum(sorted(map(sum, data), reverse=True)[:3])


data = parse_input(load_day(1, separator="\n\n"))


if __name__ == '__main__':
    print(f'Day {__day__}: {part1(data)}, {part2(data)}')

File: 2024/src/bin/day01.rs
use std::collections::HashMap;

// Process the input text, split into rows and parse them by column into the two lists.
fn process_data(contents: String) -> (Vec<u32>, Vec<u32>) {
    contents
        .lines()
        .map(|line| aoc::parse_numbers::<u32>(line))
        .map(|each| (each[0], each[1]))
        .unzip()
}

// Calculate the similarity score by calculating the distance of all the sorted pairs.
fn part_1(data: &(Vec<u32>, Vec<u32>)) -> u32 {
    let (left, right) = data;
    let mut left = left.to_vec();
    let mut right = right.to_vec();

    left.sort();
    right.sort();

    left.iter()
        .zip(right)
        .map(|(each_first, each_second)| each_first.abs_diff(each_second))
        .sum()
}

// Calculate the similarity score by multiplying each number on the left list by the number of
// they appear on the right list, then add them together.
fn part_2(data: &(Vec<u32>, Vec<u32>)) -> u32 {
    let (left, right) = data;
    // build the index with counts of the elements in the right list
    let mut counts: HashMap<u32, u32> = HashMap::new();
    for each in right {
        counts
            .entry(*each)
            .and_modify(|value| *value += 1)
            .or_insert(1);
    }

    left.iter()
        .map(|each| each * counts.get(each).unwrap_or(&0))
        .sum()
}

fn main() {
    let mut solution = aoc::Solution::day(1);
    let data = process_data(solution.data());
    solution.part1(part_1(&data)).part2(part_2(&data)).print();
}
